### ISSUE TITLE
Property injection not working in ASP.NET Core 1.0

### ISSUE THREAD

**Issue by ddotlic** [autofac/Autofac#713]
Hi,

I'm not sure what the deal is here but the property injection does not seem to work for non-controller classes in ASP.NET Core 1.0

For example, in one of the controllers we have
```
class SomeController : Controller {
    [FromServices]
    public ISomething Something { get; set; }
    // ....
}
```

This works, though I'm required to add `FromServices` (which is ASP.NET MVC's attribute) for it to work.

If I use a class not derived from `Controller`, property injection simply does not work, regardless if there's `FromServices` attribute applied or not.

On the exact same class, using constructor injection (for the same thing) works like a charm.

Can someone please explain the discrepancy and/or point me in the right direction?

Here's my reasoning for wanting to use property injection, if there's a cleaner way to achieve the same goal please do enlighten me :smile: 

We have a tiny hierarchy of classes `A` (and "many" more in the future) deriving from `B`. Most of the logic for doing things is in `B` and `B` needs two other "services" injected into it, but `A` does not care as it provides - optionally - an override of a single method or two, not using the services injected into `B` (or using indirectly but not being aware of).

If we used constructor injection, I'd be obliged - and that's what I currently do - to burden all classes like `A ` to have constructors accepting the two services in question only to pass them on to the base class.

I thought that using property injection in this case is justified, but it does not work as explained above.

Other than that, Autofac works great with the new ASP.NET platform and is a breeze to setup, thanks for your hard work!

**Comment by ddotlic** [comment]
...and of course after reading all this I went back to the documentation only to discover `PropertiesAutowired` method :wink: 

Sorry for the noise, when used properly Autofac works just fine.

**Comment by ddotlic** [comment]
OK, so this is weird: property injection works for properties "directly" on the class I'm resolving, but **not** on the base class (which is abstract  - and generic - if that has any significance, though I tried with non-abstract as well).

Based on several questions on SO, this should work (properties in question are public).

Am I missing something?

**Comment by ddotlic** [comment]
More precisely because it might not be obvious...

```
public abstract class B {
    public IProp1 Prop1 { get; set; }
    public IProp2 Prop2 { get; set; }
    // ...
}

public class A : B {
    public IProp3 Prop3 { get; set; }
    // ...
}

```

If I add `IProp1` and `IProp2` to `A` constructor (and `B` constructor of course), then pass them from `A` constructor to `B`, no problem. Property with type `IProp3` also gets injected correctly, but properties with types `IProp1` and `IProp2` do not.

I've slightly simplified classes `A` and `B` which are actually generic classes with two type parameters (so `B` is something like `B<T1,T2>` and `A` is derived from `B<ConcreteType1, ConcreteType2>`) but I believe this has no bearing on the result.

**Comment by alexmg** [comment]
This issue has already been in fixed in 0c53a056e4922c60087a7d0e7d8de8746a7f50c2 and 0c53a056e4922c60087a7d0e7d8de8746a7f50c2. When RC2 is released this fix will be included. Depending on when Microsoft releases we may do a release sooner.

**Comment by ddotlic** [comment]
Ah, great, I thought it must have been me seeing how property injection works fine elsewhere (in other scenarios Autofac is used). Thanks a lot, I'll try again as soon as RC2 is released (which should be soon-ish, in any case we can wait).

**Comment by ddotlic** [comment]
So, just tried in RC2 and for Controllers, this does not work.
The reason is likely the fact that Controllers are not registered by us, but by framework itself, so there is no line of code calling `PropertiesAutowired`.

What is the recommended approach? To register controllers "manually"? Will this not generate a confict with the framework registrations (wrapped by Autofac)?

**Comment by tillig** [comment]
As long as you manually register controllers _after_ calling `builder.Populate(services)` it should be OK.

**Comment by ddotlic** [comment]
Yeah, but that's not nearly enough (through no fault of Autofac), and the code to do so changed - significantly - between RC1 and RC2... but I figured it out :)

**Comment by pgrm** [comment]
Hi, how did you manage to get it running? I have one project with controllers and another one with the services. All services I want to be registered are marked with the interface IAutoWire. The code looks like:

    var builder = new ContainerBuilder();

    builder.Populate(services);
    builder.RegisterAssemblyTypes(Assembly.GetExecutingAssembly())
        .Where(t => typeof(Controller).IsAssignableFrom(t))
        .PropertiesAutowired();
    builder.RegisterAssemblyTypes(typeof(IAutoWire).Assembly)
        .Where(t => typeof(IAutoWire).IsAssignableFrom(t))
        .PropertiesAutowired()
        .AsImplementedInterfaces()
        .InstancePerLifetimeScope();

I tried to change the order but nothing seemed to work. I came across this article http://www.strathweb.com/2016/03/the-subtle-perils-of-controller-dependency-injection-in-asp-net-core-mvc/ pointing out the exact problem I have - the properties are still null, but the solution described doesn't work with RC2 anymore.

**Comment by ddotlic** [comment]
If you look at the comments on that page, you'll see that I provided updated code for RC2 😉

**Comment by pgrm** [comment]
@ddotlic ah sry - I just saw the first comment was 2 months ago and the last one as well - thx, will try it

**Comment by pgrm** [comment]
@ddotlic taking from your comments, this is the code I came up with:

    public IServiceProvider ConfigureServices(IServiceCollection services)
    {
        services.Replace(ServiceDescriptor.Transient<IControllerActivator, ServiceBasedControllerActivator>());
        services.AddMvc();

        var builder = new ContainerBuilder();
        var manager = new ApplicationPartManager();

        manager.ApplicationParts.Add(new AssemblyPart(Assembly.GetExecutingAssembly()));
        manager.FeatureProviders.Add(new ControllerFeatureProvider());

        var feature = new ControllerFeature();

        manager.PopulateFeature(feature);

        builder.RegisterType<ApplicationPartManager>().AsSelf().SingleInstance();
        builder.RegisterTypes(feature.Controllers.Select(ti => ti.AsType()).ToArray()).PropertiesAutowired();

        builder.RegisterAssemblyTypes(typeof(IAutoWire).Assembly)
            .Where(t => typeof(IAutoWire).IsAssignableFrom(t))
            .PropertiesAutowired()
            .AsImplementedInterfaces()
            .InstancePerLifetimeScope();

        builder.Populate(services);

        IContainer container = builder.Build();
        return container.Resolve<IServiceProvider>();
    }

but it still doesn't work. In your comment you didn't specify where and if to call `services.AddMvc();` I tried before and after calling `services.Replace(...)` but neither helped. I also tried to vary the order of `builder.RegisterTypes`, `builder.RegisterAssemblyTypes` and `builder.Populate` but none of them worked.

Constructor parameters are initialized properly all the time, so what did I miss? - Thank you very much

**Comment by ddotlic** [comment]
@pgrm In our case, call to AddMvc is done prior to all of the rest of the code. In fact, I've factored the container registration "stuff" into an extension method of its own so that in `ConfigureServices` we have "normal" ASP.NET Core stuff and the last line reads `return services.AddAutofac()`, but this is purely syntactic sugar/readability thing. In any case, call to `AddMvc` is the first one in the `ConfigureServices` (as added by the wizard when the project was initially created).

Try calling `AddMvc` first, then the rest. Verify that the `feature.Controllers` contains your controller types (if it doesn't the issue is in the code related to discovery of types in assembly "pars" though your code looks good). I honestly see nothing different from our code - after all, it's what I took from our final code and posted to the blog and you've taken, so it should work...
